src/v/pandaproxy/util.h

```diff
@@ -34,7 +34,9 @@ class one_shot {
       : _func{std::move(func)} {}
     ss::future<> operator()() {
         if (likely(_started_sem.available_units() != 0)) {
-            return _started_sem.wait();
+            // Ensure we can get a unit that the oneshot has completed
+            // then return the units.w:
+            return ss::get_units(_started_sem, 1).discard_result();
```

Comment:
I don't see that there's any difference here. `get_units(sem, 1)` calls `sem.wait(1)`, and returns an RAII folder for the units, which is immediately discarded.

It seems the result should be returned, not discarded?

So it seems https://github.com/redpanda-data/redpanda/pull/14963 should be reverted?

Reply:
this change:
wait until available unit
decrement the count
immediately increment the count

dev:
wait until available unit
decrement the count


We never increment the semaphore. I think the unit here is size_t so practically speaking we will probably never run out. But as soon as the count runs out all requests will hang because nothing increment the semaphore.

Reply:
excellent find. glancing at the original, my mind registers it as a lock with an empty critical section but it most certainly is not.